File: 动态规划/线段树优化DP/1919C. Grouping Increases.py
# https://codeforces.com/contest/1919/problem/C
# Lazy 线段树优化DP
# 1800 另有贪心做法

# 这版过不了，是python的运行效率，c++ 能过，请看下面一版
# 能过的为贪心算法，详见贪心题单

from math import inf
from typing import List
import sys

input = lambda: sys.stdin.readline().rstrip('\r\n')
print = lambda *args, end='\n', sep=' ': sys.stdout.write(
    sep.join(map(str, args)) + end
)

def II():
    return int(input())

def MII(b=0):
    return map(lambda x: int(x) - b, input().split())

def LII():
    return list(MII())

class Node:
    __slots__ = 'val', 'todo'

class LazySegmentTree:
    _TODO_INIT = 0

    def __init__(self, arr, default=0):
        if isinstance(arr, int):
            arr = [default] * arr
        n = len(arr)
        self._n = n
        self._tree = [Node() for _ in range(2 << (n - 1).bit_length())]
        self._build(arr, 1, 0, n - 1)

    def _merge_val(self, a: int, b: int) -> int:
        return min(a, b)

    def _merge_todo(self, a: int, b: int) -> int:
        return a + b

    def _apply(self, node: int, l: int, r: int, todo: int) -> None:
        cur = self._tree[node]
        cur.val += todo
        cur.todo = self._merge_todo(todo, cur.todo)

    def _spread(self, node: int, l: int, r: int) -> None:
        todo = self._tree[node].todo
        if todo == self._TODO_INIT:
            return
        m = (l + r) // 2
        self._apply(node * 2, l, m, todo)
        self._apply(node * 2 + 1, m + 1, r, todo)
        self._tree[node].todo = self._TODO_INIT

    def _maintain(self, node: int) -> None:
        self._tree[node].val = self._merge_val(self._tree[node * 2].val, self._tree[node * 2 + 1].val)

    def _build(self, a: List[int], node: int, l: int, r: int) -> None:
        self._tree[node].todo = self._TODO_INIT
        if l == r:
            self._tree[node].val = a[l]
            return
        m = (l + r) // 2
        self._build(a, node * 2, l, m)
        self._build(a, node * 2 + 1, m + 1, r)
        self._maintain(node)

    def _update(self, node: int, l: int, r: int, ql: int, qr: int, f: int) -> None:
        if ql <= l and r <= qr:
            self._apply(node, l, r, f)
            return
        self._spread(node, l, r)
        m = (l + r) // 2
        if ql <= m:
            self._update(node * 2, l, m, ql, qr, f)
        if qr > m:
            self._update(node * 2 + 1, m + 1, r, ql, qr, f)
        self._maintain(node)

    def _query(self, node: int, l: int, r: int, ql: int, qr: int) -> int:
        if ql <= l and r <= qr:
            return self._tree[node].val
        self._spread(node, l, r)
        m = (l + r) // 2
        if qr <= m:
            return self._query(node * 2, l, m, ql, qr)
        if ql > m:
            return self._query(node * 2 + 1, m + 1, r, ql, qr)
        l_res = self._query(node * 2, l, m, ql, qr)
        r_res = self._query(node * 2 + 1, m + 1, r, ql, qr)
        return self._merge_val(l_res, r_res)

    def update(self, ql: int, qr: int, f: int) -> None:
        self._update(1, 0, self._n - 1, ql, qr, f)

    def query(self, ql: int, qr: int) -> int:
        return self._query(1, 0, self._n - 1, ql, qr)
    
    def get(self, i):
        return self.query(i, i)
    
    def set_val(self, i, new):
        cur = self.get(i)
        self.update(i, i, new - cur)

def discretize(arr):
    sorted_unique = sorted(set(arr))
    value_to_index = {value: idx for idx, value in enumerate(sorted_unique)}
    discretized = [value_to_index[x] for x in arr]
    return discretized, len(sorted_unique)

def solve():
    n = II()
    nums = LII()
    
    if n <= 1:
        print(0)
        return
    
    nums, m = discretize(nums)
    
    INF = 10**9
    seg = LazySegmentTree(m + 1, INF)
    seg.set_val(m, 0)
    
    for i in range(1, n):
        val1 = seg.query(0, nums[i] - 1) + 1 if nums[i] > 0 else INF
        val2 = seg.query(nums[i], m)
        ndp = min(val1, val2)

        if nums[i - 1] < nums[i]:
            seg.update(0, m, 1)

        current_val = seg.get(nums[i - 1])
        if ndp < current_val:
            seg.set_val(nums[i - 1], ndp)
    
    print(seg.query(0, m))

if __name__ == "__main__":
    for _ in range(II()):
        solve()

File: DP/DP_Optimization/SegmentTree_Optimization/1919C_Grouping_Increases.cpp
// https://codeforces.com/contest/1919/problem/C
// Lazy 线段树优化DP
// 1800（估计的，感觉低了） 另有贪心做法

// c++ 版本，依旧是灵神的板子，能过
// 另有贪心算法，详见贪心题单


#include <bits/stdc++.h>
using namespace std;

using ll = long long;
const int INF = 1e9;

// 线段树节点类型
struct NodeVal {
    int val;
};

// 懒标记类型
struct LazyTag {
    int add;
    LazyTag(int a = 0) : add(a) {}
};

class LazySegmentTree {
private:
    vector<NodeVal> tree;
    vector<LazyTag> lazy;
    int n;
    
    // 合并两个节点值
    NodeVal merge(const NodeVal& a, const NodeVal& b) {
        return {min(a.val, b.val)};
    }
    
    // 应用懒标记
    void apply(int node, int l, int r, const LazyTag& tag) {
        tree[node].val += tag.add;
        lazy[node].add += tag.add;
    }
    
    // 下传懒标记
    void push(int node, int l, int r) {
        if (lazy[node].add == 0) return;
        int mid = (l + r) >> 1;
        apply(node << 1, l, mid, lazy[node]);
        apply(node << 1 | 1, mid + 1, r, lazy[node]);
        lazy[node].add = 0;
    }
    
    void build(int node, int l, int r, int init_val) {
        if (l == r) {
            tree[node] = {init_val};
            return;
        }
        int mid = (l + r) >> 1;
        build(node << 1, l, mid, init_val);
        build(node << 1 | 1, mid + 1, r, init_val);
        tree[node] = merge(tree[node << 1], tree[node << 1 | 1]);
    }
    
    void update_range(int node, int l, int r, int ql, int qr, int val) {
        if (ql <= l && r <= qr) {
            tree[node].val += val;
            lazy[node].add += val;
            return;
        }
        push(node, l, r);
        int mid = (l + r) >> 1;
        if (ql <= mid) update_range(node << 1, l, mid, ql, qr, val);
        if (qr > mid) update_range(node << 1 | 1, mid + 1, r, ql, qr, val);
        tree[node] = merge(tree[node << 1], tree[node << 1 | 1]);
    }
    
    void update_point(int node, int l, int r, int idx, int val) {
        if (l == r) {
            tree[node].val = val;
            return;
        }
        push(node, l, r);
        int mid = (l + r) >> 1;
        if (idx <= mid) update_point(node << 1, l, mid, idx, val);
        else update_point(node << 1 | 1, mid + 1, r, idx, val);
        tree[node] = merge(tree[node << 1], tree[node << 1 | 1]);
    }
    
    NodeVal query(int node, int l, int r, int ql, int qr) {
        if (ql <= l && r <= qr) return tree[node];
        push(node, l, r);
        int mid = (l + r) >> 1;
        if (qr <= mid) return query(node << 1, l, mid, ql, qr);
        if (ql > mid) return query(node << 1 | 1, mid + 1, r, ql, qr);
        return merge(query(node << 1, l, mid, ql, qr),
                     query(node << 1 | 1, mid + 1, r, ql, qr));
    }
    
public:
    LazySegmentTree(int size, int init_val) : n(size) {
        tree.resize(4 * n);
        lazy.resize(4 * n);
        build(1, 0, n - 1, init_val);
    }
    
    void update(int l, int r, int val) {
        update_range(1, 0, n - 1, l, r, val);
    }
    
    void set(int idx, int val) {
        update_point(1, 0, n - 1, idx, val);
    }
    
    int query(int l, int r) {
        return query(1, 0, n - 1, l, r).val;
    }
    
    int get(int idx) {
        return query(idx, idx);
    }
};

vector<int> discretize(vector<int>& arr, int& m) {
    vector<int> sorted_unique = arr;
    sort(sorted_unique.begin(), sorted_unique.end());
    sorted_unique.erase(unique(sorted_unique.begin(), sorted_unique.end()), sorted_unique.end());
    m = sorted_unique.size();
    vector<int> discretized;
    for (int x : arr) {
        discretized.push_back(lower_bound(sorted_unique.begin(), sorted_unique.end(), x) - sorted_unique.begin());
    }
    return discretized;
}

void solve() {
    int n;
    cin >> n;
    vector<int> nums(n);
    for (int i = 0; i < n; i++) {
        cin >> nums[i];
    }
    
    if (n <= 1) {
        cout << 0 << '\n';
        return;
    }
    
    int m;
    auto discretized_nums = discretize(nums, m);
    
    LazySegmentTree seg(m + 1, INF);
    seg.set(m, 0);
    
    for (int i = 1; i < n; i++) {
        int val1 = (discretized_nums[i] > 0) ? seg.query(0, discretized_nums[i] - 1) + 1 : INF;
        int val2 = seg.query(discretized_nums[i], m);
        int ndp = min(val1, val2);
        
        if (discretized_nums[i - 1] < discretized_nums[i]) {
            seg.update(0, m, 1);
        }
        
        int current_val = seg.get(discretized_nums[i - 1]);
        if (ndp < current_val) {
            seg.set(discretized_nums[i - 1], ndp);
        }
    }
    
    cout << seg.query(0, m) << '\n';
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    
    int T;
    cin >> T;
    
    while (T--) {
        solve();
    }
    
    return 0;
}